gh-125038: Iterator checks are added for some FOR_ITER bytecodes, crash fixed

efimov-mikhail opened the pull request:
SIGSEGV on generators in case of gi_frame.f_locals is fixed.

This applies to _FOR_ITER bytecode implementation.
Similar checks are added to _FOR_ITER_TIER_TWO and INSTRUMENTED_FOR_ITER bytecode implementations.

<!-- gh-issue-number: gh-125038 -->
* Issue: gh-125038
<!-- /gh-issue-number -->

efimov-mikhail committed:
gh-125038: PyIter_Checks are added for some FOR_ITER bytecodes

SIGSEGV on generators in case of gi_frame.f_locals is fixed.
This applies to _FOR_ITER bytecode.
Similar checks are added to _FOR_ITER_TIER_TWO and INSTRUMENTED_FOR_ITER bytecode implementations.

efimov-mikhail committed:
gh-125038: Test on gi_frame.f_locals change for a generator

efimov-mikhail committed:
gh-125038: News entry is added.

JelleZijlstra commented on Python/bytecodes.c:
```diff
@@ -2804,7 +2804,10 @@ dummy_func(
             PyObject *iter_o = PyStackRef_AsPyObjectBorrow(iter);
-            PyObject *next_o = (*Py_TYPE(iter_o)->tp_iternext)(iter_o);
+            PyObject *next_o = NULL;
+            if (PyIter_Check(iter_o)) {
```
Calling `PyIter_Check` feels a little wasteful. Instead, I'd assign `(*Py_TYPE(iter_o)->tp_iternext)` to a variable and check if it's NULL. If it is NULL, we should raise a TypeError. Your code instead makes it immediately end the loop, which doesn't feel like the right behavior.

efimov-mikhail replied:
Hmm, I've tried to add a TypeError 
Such piece of code is provided:

```
            PyTypeObject *type = Py_TYPE(iter_o);
            iternextfunc iternext = type->tp_iternext;
            if (iternext == NULL) {
                _PyErr_Format(tstate, PyExc_TypeError,
                              "'for' requires an object with "
                              "__iter__ method, got %.100s",
                              type->tp_name);
                DECREF_INPUTS();
                ERROR_IF(true, error);
            }
            PyObject *next_o = (*iternext)(iter_o);
```

And there is a temporarily result of my test case:

```
-> % ./python -m unittest -v test.test_generators.GeneratorTest.test_issue125038
test_issue125038 (test.test_generators.GeneratorTest.test_issue125038) ... ERROR

======================================================================
ERROR: test_issue125038 (test.test_generators.GeneratorTest.test_issue125038)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mikhail.efimov/projects/cpython/Lib/test/test_generators.py", line 274, in test_issue125038
    l = list(g)
  File "/home/mikhail.efimov/projects/cpython/Lib/test/test_generators.py", line 272, in <genexpr>
    g = (x for x in range(10))
                    ~~~~~^^^^
TypeError: 'for' requires an object with __iter__ method, got range

----------------------------------------------------------------------
Ran 1 test in 0.001s
```

It seems like such a message doesn't provide any clarity.
Do you have any better suggestions about error message?

JelleZijlstra replied:
That seems like a pretty good error message, much better than silently stopping the loop.

efimov-mikhail committed:
gh-125038: Raise TypeError on such case instead of silently stopping the loop

JelleZijlstra commented on Lib/test/test_generators.py:
```diff
@@ -269,10 +269,22 @@ def loop():
+            except TypeError:
+                return "TypeError"
+
+        # This should not raise
```
I think you want `with self.assertRaisesRegex`.

efimov-mikhail committed:
gh-125038: Test improvement, assertRaisesRegex is used

vstinner committed:
gh-124502: Add PyUnicode_Equal() function (#124504)

emilyemorehouse committed:
gh-125072: Add label for assignment expressions; update tracked section for assignment expression topic (#125074)

nedbat committed:
Docs: make a tutorial example more precise (#125066)

Based on discussion here:

https://discuss.python.org/t/omission-in-the-documentation/66816

JelleZijlstra commented on Misc/NEWS.d/next/Core_and_Builtins/2024-10-07-19-26-50.gh-issue-125038.ffSLCz.rst:
```diff
@@ -0,0 +1,2 @@
+Fix SIGSEGV in genexpr after direct changes on gi_frame.f_locals. Patch by
```
```suggestion
Fix crash when iterating over a generator expression after direct changes on ``gi_frame.f_locals``. Patch by
```
The exact crash mechanism is OS-dependent and not guaranteed, so let's just say "crash".

JelleZijlstra reviewed (commented):
This looks correct to me, but will leave to Mark to see if there's a better solution.

efimov-mikhail committed:
Update Misc/NEWS.d/next/Core_and_Builtins/2024-10-07-19-26-50.gh-issue-125038.ffSLCz.rst

News improvement

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

cmaloney committed:
gh-90102: Remove isatty call during regular open (#124922)

Co-authored-by: Victor Stinner <vstinner@python.org>

ncoghlan committed:
gh-70870: Clarify dual usage of 'free variable' (#122545)

The term "free variable" has unfortunately become genuinely
ambiguous over the years (presumably due to the names of
some relevant code object instance attributes).

While we can't eliminate that ambiguity at this late date, we can
at least alert people to the potential ambiguity by describing
both the formal meaning of the term and the common
alternative use as a direct synonym for "closure variable".

---------

Co-authored-by: Carol Willing <carolcode@willingconsulting.com>

serhiy-storchaka committed:
gh-69998: Fix decoding error in locale.nl_langinfo() (GH-124963)

The function now sets temporarily the LC_CTYPE locale to the locale
of the category that determines the requested value if the locales are
different and the resulting string is non-ASCII.
This temporary change affects other threads.

ncoghlan committed:
bpo-34206: Improve docs and test coverage for pre-init functions (#8023)

- move the Py_Main documentation from the very high level API section
  to the initialization and finalization section
- make it clear that it encapsulates a full Py_Initialize/Finalize
  cycle of its own
- point out that exactly which settings will be read and applied
  correctly when Py_Main is called after a separate runtime
  initialization call is version dependent
- be explicit that Py_IsInitialized can be called prior to
  initialization
- actually test that Py_IsInitialized can be called prior to
  initialization
- flush stdout in the embedding tests that run code so it appears
  in the expected order when running with "-vv"
- make "-vv" on the subinterpreter embedding tests less spammy

---------

Co-authored-by: Carol Willing <carolcode@willingconsulting.com>

serhiy-storchaka committed:
gh-53203: Improve tests for strptime() (GH-125090)

Run them with different locales and different date and time.

Add the @run_with_locales() decorator to run the test with multiple
locales.

Improve the run_with_locale() context manager/decorator -- it now
catches only expected exceptions and reports the test as skipped if no
appropriate locale is available.

picnixz committed:
gh-123378: fix a crash in `UnicodeError.__str__` (#124935)

picnixz committed:
gh-123961: Convert _curses to a multi-phase init module (PEP-489) (#124965)

cmaloney committed:
gh-90102: Fix pyio _isatty_open_only() (#125089)

Spotted by @ngnpope.

`isatty` returns False to indicate the file is not a TTY. The C
implementation of _io does that (`Py_RETURN_FALSE`) but I got the
bool backwards in the _pyio implementaiton.

zuo committed:
gh-75898: make use of thread more explicit in the "Socket Programming HOWTO" document (#125023)

efimov-mikhail committed:
gh-121404: typo fix in compile.c: MATADATA -> METADATA (#125101)

Eclips4 committed:
Doc: Improve description of ``GET_LEN`` opcode (#114583)

monkeyman192 committed:
gh-112433 add versionadded for `ctypes.Structure._align_` (#125087)


Co-authored-by: Kumar Aditya <kumaraditya@python.org>

vstinner committed:
gh-125096: Don't import _pyrepl in site if PYTHON_BASIC_REPL (#125097)

If the PYTHON_BASIC_REPL environment variable is set, the site module
no longer imports the _pyrepl module.

Moreover, the site module now respects -E and -I command line
options: ignore PYTHON_BASIC_REPL in this case.

mpage committed:
gh-115999: Stop the world when invalidating function versions (#124997)

Stop the world when invalidating function versions

The tier1 interpreter specializes `CALL` instructions based on the values
of certain function attributes (e.g. `__code__`, `__defaults__`). The tier1
interpreter uses function versions to verify that the attributes of a function
during execution of a specialization match those seen during specialization.
A function's version is initialized in `MAKE_FUNCTION` and is invalidated when
any of the critical function attributes are changed. The tier1 interpreter stores
the function version in the inline cache during specialization. A guard is used by
the specialized instruction to verify that the version of the function on the operand
stack matches the cached version (and therefore has all of the expected attributes).
It is assumed that once the guard passes, all attributes will remain unchanged
while executing the rest of the specialized instruction.

Stopping the world when invalidating function versions ensures that all critical
function attributes will remain unchanged after the function version guard passes
in free-threaded builds. It's important to note that this is only true if the remainder
of the specialized instruction does not enter and exit a stop-the-world point.

We will stop the world the first time any of the following function attributes
are mutated:

- defaults
- vectorcall
- kwdefaults
- closure
- code

This should happen rarely and only happens once per function, so the performance
impact on majority of code should be minimal.

Additionally, refactor the API for manipulating function versions to more clearly
match the stated semantics.

vstinner committed:
gh-124502: Optimize unicode_eq() (#125105)

kumaraditya303 committed:
gh-117721: use PyMutex in `_thread.lock` (#125110)

cmaloney committed:
gh-125084: Resolve paths in generator common code (GH-125085)

In out of tree builds, the paths can contain `../ which needs to be
resolved for the relative path calculation to work.

mdboom committed:
gh-125063: Emit slices as constants in the bytecode compiler (#125064)

* Make slices marshallable

* Emit slices as constants

* Update Python/marshal.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

* Refactor codegen_slice into two functions so it
always has the same net effect

* Fix for free-threaded builds

* Simplify marshal loading of slices

* Only return SUCCESS/ERROR from codegen_slice

---------

Co-authored-by: Mark Shannon <mark@hotpy.org>
Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

spacemanspiff2007 committed:
gh-124832: Add a note to indicate that `datetime.now` may return the same instant (#124834)

* Update datetime.rst

* Update datetime.rst

replace warning with note

* Update Doc/library/datetime.rst

Co-authored-by: Victor Stinner <vstinner@python.org>

* Update Doc/library/datetime.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

---------

Co-authored-by: Victor Stinner <vstinner@python.org>
Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

rhettinger committed:
Misc improvements to the itertools docs (gh-125147)

skirpichev committed:
gh-116110: remove extra processing for the __signature__ attribute (GH-116234)

This is an alternative to GH-100168.

savannahostrowski committed:
GH-124478: Cleanup argparse documentation (#124877)

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>
Co-authored-by: Tomas R <tomas.roun8@gmail.com>

felixxm committed:
gh-123849: Fix test_sqlite3.test_table_dump when foreign keys are enabled by default (#123859)

corona10 committed:
gh-124612: Use ghcr.io/python/autoconf instead of public image (#124657)

* gh-124612: Use ghcr.io/python/autoconf instead of public image

* Update

JelleZijlstra committed:
gh-101552: Allow pydoc to display signatures in source format (#124669)

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

musvaage committed:
docs: in venv table use executable name (GH-124315)

Co-authored-by: musvaage <musvaage@users.noreply.github.com>

vstinner committed:
gh-124502: Remove _PyUnicode_EQ() function (#125114)

* Replace unicode_compare_eq() with unicode_eq().
* Use unicode_eq() in setobject.c.
* Replace _PyUnicode_EQ() with _PyUnicode_Equal().
* Remove unicode_compare_eq() and _PyUnicode_EQ().

furkanonder committed:
gh-125150: Skip test_fma_zero_result on NetBSD due to IEE 754-2008 implementation issues (#125151)

Skip test_fma_zero_result on NetBSD due to IEE 754-2008 implementation issues

serhiy-storchaka committed:
gh-124969: Fix locale.nl_langinfo(locale.ALT_DIGITS) (GH-124974)

Now it returns a tuple of up to 100 strings (an empty tuple on most locales).
Previously it returned the first item of that tuple or an empty string.

efimov-mikhail committed:
gh-125038: Tests on generator modifying through gi_frame.f_locals are added

efimov-mikhail committed:
Merge branch 'sigsegv_fix_in_iterators' of github.com:efimov-mikhail/cpython into sigsegv_fix_in_iterators

kanavin committed:
gh-107562: make_ssl_certs.py: produce test certificates that expire far in the future by default (GH-107594)

This allows testing Y2038 with system time set to after that,
so that actual Y2038 issues can be exposed, and not masked
by expired certificate errors.

Signed-off-by: Alexander Kanavin <alex.kanavin@gmail.com>

efimov-mikhail committed:
gh-125038: Little change in test, readability slightly improved
